pulp_ansible/app/galaxy/v3/viewsets.py

```diff
@@ -0,0 +1,69 @@
+from django.db.models import Value, F, CharField
+from django.db.models import When, Case
+from django.db.models import Q
+from rest_framework.pagination import PageNumberPagination
+from django_filters.rest_framework import DjangoFilterBackend
+from django.db.models.functions import Concat
+from django_filters import filters
+from rest_framework import viewsets
+
+from pulp_ansible.app.galaxy.v3.serializers import (
+    CollectionVersionSearchListSerializer,
+    CollectionVersionListSerializer
+)
+
+from pulpcore.app.models import RepositoryContent
+
+from pulp_ansible.app.models import (
+    AnsibleCollectionDeprecated,
+    AnsibleDistribution,
+    CollectionVersion,
+)
+
+from pulp_ansible.app.viewsets import (
+    CollectionVersionFilter,
+)
+
+
+class CollectionVersionSearchViewSetPagination(PageNumberPagination):
+    page_size = 100
+    page_size_query_param = "page_size"
+    max_page_size = 1000
+
+
+class CollectionVersionSearchViewSet(viewsets.ModelViewSet):
+
+    serializer_class = CollectionVersionSearchListSerializer
+    pagination_class = CollectionVersionSearchViewSetPagination
+    filter_backends = (DjangoFilterBackend,)
+    filterset_class = None
+
+    def get_queryset(self):
+
+        include_q = Q()
+        exclude_q = Q()
+
+        for distro in AnsibleDistribution.objects.all():
+            print(f'DISTRO: {distro} {distro.name}')
+            if not distro.repository_version and distro.repository is None:
+                continue
+            elif distro.repository_version:
+                rv = distro.repository_version.number
+            else:
+                rv = distro.repository.latest_version().number
+
+            include_q = include_q | Q(repository=distro.repository, version_added__number__lte=rv)
+            exclude_q = exclude_q | Q(repository=distro.repository, version_removed__number__lte=rv)
+
+        qs = RepositoryContent.objects.filter(
+                content__pulp_type="ansible.collection_version"
+            ).exclude(
+                exclude_q
+            ).filter(
+                include_q
+            )
```

Comment:
This logic makes sense for trying to keep the query count to 1, but have you considered using RepositoryVersion's `get_content` method? https://github.com/pulp/pulpcore/blob/main/pulpcore/app/models/repository.py#L654. With this method you can pass in an existing CollectionVersion queryset to avoid having to cast the objects in the serializer. You would have to OR together the queries across each distribution's repository's content, but it technically should only result in one query to the DB as well, however I'm not sure if it would be as performant as your way now.

Reply:
My one worry with your current approach is that each `cast` call results in a db query, so as the number of results for a search increases, so does the number of queries required to show the results.

Reply:
Is .content.cast() the only way to access the CollectionVersion attributes?

Reply:
Yes. It probably makes more sense to convert the RepositoryContent query to a CollectionVersion query to avoid casting for each result.